src/cli/telemetry/telemetry.ts

```diff
@@ -0,0 +1,293 @@
+import { CompilerBuildResults, CompilerSystem, Config, Logger, PackageJsonData } from 'src/declarations';
+import { parseFlags } from '../parse-flags';
+import { loadCoreCompiler } from '../load-compiler';
+import { readConfig, updateConfig } from '../ionic-config';
+import { isObject } from '@utils';
+import { getCompilerSystem } from '../state/stencil-cli-config';
+
+export interface CommandMetricData {
+  // app_id: string;
+  command: string;
+  arguments: string;
+  options: string;
+  duration: number;
+  error: string | null;
+  node_version: string;
+  // os: string;
+}
+
+export interface Metric<N extends string, D> {
+  name: N;
+  timestamp: string;
+  source: 'stencil_cli';
+  value: D;
+}
+
+// Used for deeper understanding on how people are using certain output targets
+interface OutputTargetOptions {
+  serviceWorker?: boolean;
+}
+
+type TelemetryCallback = (...args: any[]) => void | Promise<void>;
+
+export interface TrackableData {
+  command?: string;
+  yarn?: boolean;
+  session_id?: string;
+  component_count?: number;
+  arguments?: string[];
+  targets?: string[][] | OutputTargetOptions[][];
+  task?: string;
+  duration?: number;
+  packages?: string[];
+  telemetry?: boolean;
+  os_name: string;
+  os_version: string;
+  cpu_model: string;
+  typescript: string;
+  rollup: string;
+  system: string;
+  build: string;
+  stencil: string;
+}
+
+export interface telemetryActionArgs {
+  args?: string[];
+  sys?: CompilerSystem;
+  logger?: Logger;
+  buildResults?: Logger;
+  config?: Config;
+  action?: TelemetryCallback;
+}
+
+// Used to within taskBuild to signal out with the componentCount integer.
+export async function telemetryBuildFinishedAction(result: CompilerBuildResults, args: telemetryActionArgs) {
+  const { sys, config, logger } = args;
+  const tracking = await shouldTrack(!!config.flags.ci);
+
+  if (!tracking) {
+    return;
+  }
+
+  const details = sys.details;
+  const session_id = await getTelemetryToken();
+  const packages = await getInstalledPackages();
+  const targets = await getActiveTargets(config);
+  const coreCompiler = await loadCoreCompiler(sys);
+  const versions = coreCompiler.versions;
+  const component_count = Object.keys(result.componentGraph).length;
+
+  const data: TrackableData = {
+    session_id,
+    yarn: isUsingYarn(),
+    duration: result.duration,
+    component_count,
+    targets,
+    packages: packages || [],
+    arguments: config.flags.args,
+    task: config.flags.task,
+    stencil: coreCompiler.version,
+    system: `${sys.name} ${sys.version}`,
+    os_name: details.platform,
+    os_version: details.release,
+    cpu_model: `${details.cpuModel}`,
+    build: `${result.buildId}`,
+    typescript: versions.typescript,
+    rollup: versions.rollup,
+  };
+
+  await sendMetric('stencil_cli_command', data);
+  logger.debug(`${logger.blue('Telemetry')}: ${logger.gray(JSON.stringify(data))}`);
+}
+
+export async function telemetryAction(options: telemetryActionArgs) {
```

Comment:
Can we get these publicly exposed methods under test please?

Reply:
Working on this rn .